server/leader.go

```diff
@@ -16,234 +16,118 @@ package server
-	resp, err := s.txn().
-		If(clientv3.Compare(clientv3.CreateRevision(leaderKey), "=", 0)).
-		Then(clientv3.OpPut(leaderKey, s.leaderValue, clientv3.WithLease(clientv3.LeaseID(leaseResp.ID)))).
-		Commit()
-	if err != nil {
-		return errors.Trace(err)
-	}
-	if !resp.Succeeded {
-		return errors.New("campaign leader failed, other server may campaign ok")
+		log.Infof("campaign leader ok: %v", s.Name())
+		s.leaderRound(lessor)
 	}
+}
 
-	log.Debugf("campaign leader ok %s", s.Name())
-	s.enableLeader(true)
-	defer s.enableLeader(false)
+func (s *Server) leaderRound(lessor *Lessor) {
+	s.becomeLeader(lessor)
+	defer s.resignLeader()
 
-	// Try to create raft cluster.
-	err = s.createRaftCluster()
-	if err != nil {
-		return errors.Trace(err)
+	if err := s.createRaftCluster(); err != nil {
+		log.Error(err)
+		return
 	}
 
-	// Make the leader keepalived.
-	ch, err := lessor.KeepAlive(s.client.Ctx(), clientv3.LeaseID(leaseResp.ID))
-	if err != nil {
-		return errors.Trace(err)
+	if err := s.syncTimestamp(); err != nil {
+		log.Error(err)
+		return
 	}
 
-	log.Debug("sync timestamp for tso")
-	if err = s.syncTimestamp(); err != nil {
-		return errors.Trace(err)
-	}
-
-	tsTicker := time.NewTicker(updateTimestampStep)
-	defer tsTicker.Stop()
+	ticker := time.NewTicker(updateTimestampStep)
+	defer ticker.Stop()
 
 	for {
 		select {
-		case _, ok := <-ch:
-			if !ok {
-				log.Info("keep alive channel is closed")
-				return nil
-			}
-		case <-tsTicker.C:
-			if err = s.updateTimestamp(); err != nil {
-				return errors.Trace(err)
+		case <-ticker.C:
+			if err := s.updateTimestamp(); err != nil {
+				log.Error(err)
+				return
 			}
-		case <-s.client.Ctx().Done():
-			return errors.New("server closed")
+		case <-lessor.Done():
+			return
 		}
 	}
 }
 
-func (s *Server) watchLeader() {
-	watcher := clientv3.NewWatcher(s.client)
-	defer watcher.Close()
-
-	ctx := s.client.Ctx()
-	for {
-		rch := watcher.Watch(ctx, s.getLeaderPath())
-		for wresp := range rch {
-			if wresp.Canceled {
-				return
-			}
-
-			for _, ev := range wresp.Events {
-				if ev.Type == mvccpb.DELETE {
-					log.Info("leader is deleted")
-					return
-				}
-			}
-		}
+func (s *Server) becomeLeader(lessor *Lessor) {
+	s.lessor.Store(lessor)
+	s.closeAllConnections()
+}
 
-		select {
-		case <-ctx.Done():
-			// server closed, return
-			return
-		default:
-		}
+func (s *Server) resignLeader() {
+	lessor := s.getLessor()
+	if lessor != nil {
+		lessor.Close()
 	}
+	s.lessor.Store((*Lessor)(nil))
+	s.closeAllConnections()
+	s.cluster.stop()
 }
 
-func (s *Server) resignLeader() error {
-	// delete leader itself and let others start a new election again.
-	leaderKey := s.getLeaderPath()
-	resp, err := s.leaderTxn().Then(clientv3.OpDelete(leaderKey)).Commit()
-	if err != nil {
-		return errors.Trace(err)
-	}
-	if !resp.Succeeded {
-		return errors.New("resign leader failed, we are not leader already")
+func (s *Server) getLessor() *Lessor {
+	lessor := s.lessor.Load()
```

Comment:
seem that you can use `lessor, _ := s.lessor.Load().(*Lessor)` and return lessor directly.